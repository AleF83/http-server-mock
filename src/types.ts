import { v4 as uuidV4 } from 'uuid';

export interface ServiceMockOptions {
  name: string;
  port: number;
}

export interface ServiceMockInfo {
  id: string;
  name: string;
  port: number;
}

export class ServiceMock {
  private _id: string;
  private _name: string;
  private _port: number;

  constructor(options: ServiceMockOptions) {
    this._id = uuidV4();
    this._name = options.name;
    this._port = options.port;
  }

  get id(): string {
    return this._id;
  }

  get name(): string {
    return this._name;
  }

  get port(): number {
    return this._port;
  }

  get info(): ServiceMockInfo {
    return {
      id: this.id,
      name: this.name,
      port: this.port,
    };
  }
}
