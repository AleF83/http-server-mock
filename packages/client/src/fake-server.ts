import { AxiosInstance } from 'axios';
import { RegisterResponseMockRequest, RegisterResponseMockResponse, FakeServerInfo } from 'http-server-mock-common';

export class FakeServerClient {
  constructor(private readonly _id: string, private readonly _axios: AxiosInstance) {}

  get id(): string {
    return this._id;
  }

  async start(): Promise<void> {
    const response = await this._axios.post<void>(`/servers/${this._id}/start`);
    return response.data;
  }

  async stop(): Promise<void> {
    const response = await this._axios.post<void>(`/servers/${this._id}/stop`);
    return response.data;
  }

  async getInfo(): Promise<FakeServerInfo> {
    const response = await this._axios.get<FakeServerInfo>(`/servers/${this._id}`);
    return response.data;
  }

  async registerResponseMock(responseMock: RegisterResponseMockRequest): Promise<RegisterResponseMockResponse> {
    const response = await this._axios.post<RegisterResponseMockResponse>(
      `/servers/${this._id}/register-response-mock`,
      responseMock
    );
    return response.data;
  }
}
