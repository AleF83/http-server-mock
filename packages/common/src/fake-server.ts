import { RequestMatcher } from './request-matching';

export enum FakeServerStatus {
  Created = 'created',
  Initialized = 'initialized',
  Starting = 'starting',
  Running = 'running',
  Stopping = 'stopping',
  Stopped = 'stopped',
}

export interface FakeServerInfo {
  id: string;
  name: string;
  port: number;
  status: FakeServerStatus;
}

export interface CreateFakeServerRequest {
  name: string;
  port: number;
  startOnInit?: boolean;
}

export interface CreateFakeServerResponse {
  id: string;
  status: FakeServerStatus;
}

export interface ResponseExpectation<T = any> {
  headers?: { [key: string]: string };
  body?: T;
}

export interface ResponseBehavior {
  delay: number;
}

export interface RegisterResponseMockRequest {
  requestMatcher?: RequestMatcher;
  expectedResponse?: ResponseExpectation;
  behavior?: ResponseBehavior;
}

export interface RegisterResponseMockResponse {
  id: string;
}
