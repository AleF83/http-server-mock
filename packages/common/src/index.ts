export * from './request-matching';
export * from './fake-server';
export * from './server-calls';
