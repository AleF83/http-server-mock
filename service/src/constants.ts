export const ERR_MOCK_PORT_ALREADY_IN_USE = 'The port is already in use';
export const ERR_MOCK_ID_DOES_NOT_EXIST = "The mock with giver id doesn't exist";
