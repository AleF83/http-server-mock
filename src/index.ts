import startServer from './server';

const server = startServer();

server.listen(8080, (err, address) => {
  if (err) {
    console.error(err);
    process.exit(1);
  }
  console.log(`Server listening at ${address}`);
});
