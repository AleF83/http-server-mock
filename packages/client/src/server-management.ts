import { AxiosInstance } from 'axios';
import { CreateFakeServerRequest, CreateFakeServerResponse, FakeServerInfo } from 'http-server-mock-common';
import { FakeServerClient } from './fake-server';

export class ServerManagementClient {
  constructor(private readonly _axios: AxiosInstance) {}

  async createFakeServer(request: CreateFakeServerRequest): Promise<FakeServerClient> {
    const response = await this._axios.post<CreateFakeServerResponse>('', request);
    const fakeServerId = response.data.id;
    return new FakeServerClient(fakeServerId, this._axios);
  }

  async deleteFakeServer(id: string): Promise<void> {
    await this._axios.delete<void>(`/${id}`);
  }

  async listFakeServers(): Promise<FakeServerInfo[]> {
    const response = await this._axios.get<FakeServerInfo[]>('');
    return response.data;
  }

  async getFakeServerClient(id: string): Promise<FakeServerClient> {
    const response = await this._axios.get<FakeServerInfo>(`/${id}`);
    return new FakeServerClient(response.data.id, this._axios);
  }
}
