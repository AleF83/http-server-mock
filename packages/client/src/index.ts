import axios from 'axios';
import { ServerManagementClient } from './server-management';
export * from './server-management';
export * from './fake-server';

const DEFAULT_SERVER_MANAGER_PORT = '8080';

export function createServerManagementClient(baseUrl: string): ServerManagementClient {
  const url = new URL(baseUrl);
  if (!url.port) {
    url.port = DEFAULT_SERVER_MANAGER_PORT;
  }
  const axiosInstance = axios.create({ baseURL: `${url.origin}/servers` });
  return new ServerManagementClient(axiosInstance);
}
