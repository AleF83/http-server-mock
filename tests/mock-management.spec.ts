import { Test, TestingModule } from '@nestjs/testing';
import { FastifyAdapter, NestFastifyApplication } from '@nestjs/platform-fastify';
import { v4 as uuidV4 } from 'uuid';
import * as Constants from '../src/constants';
import { AppModule } from '../src/app.module';
import { MockManagementService } from '../src/mock-management/mock-management.service';
import { ServiceMock, ServiceMockOptions } from '../src/types';

const createTestServiceMockOptions = (): ServiceMockOptions => ({
  name: 'my service mock',
  port: 3333,
});

const createTestServiceMock = (): ServiceMock => new ServiceMock(createTestServiceMockOptions());

describe('Mock Management', () => {
  let app: NestFastifyApplication;

  beforeAll(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = moduleFixture.createNestApplication(new FastifyAdapter(), { logger: true });

    await app.init();
  });

  afterAll(async () => {
    await app.close();
  });

  test('Get all mocks', async () => {
    // Arrange
    const serviceMock = createTestServiceMock();
    const mockManagementService = app.get(MockManagementService);
    jest.spyOn(mockManagementService, 'getMocks').mockReturnValueOnce([serviceMock]);

    // Act
    const response = await app.inject({
      url: '/',
      method: 'GET',
    });

    // Assert
    expect(response.statusCode).toEqual(200);
    expect(response.json()).toEqual([serviceMock.info]);
  });

  test('Get mock by id', async () => {
    // Arrange
    const mockManagementService = app.get(MockManagementService);
    const mockService = createTestServiceMock();
    jest.spyOn(mockManagementService, 'getMock').mockReturnValueOnce(mockService);

    // Act
    const response = await app.inject({
      url: `/${mockService.id}`,
      method: 'GET',
    });

    // Assert
    expect(response.statusCode).toEqual(200);
    expect(response.json()).toEqual(mockService.info);
  });

  test('Get mock by nonexistent id', async () => {
    // Arrange
    const mockId = uuidV4();
    const mockManagementService = app.get(MockManagementService);
    // eslint-disable-next-line unicorn/no-useless-undefined
    jest.spyOn(mockManagementService, 'getMock').mockReturnValueOnce(undefined);

    // Act
    const response = await app.inject({
      url: `/${mockId}`,
      method: 'GET',
    });

    // Assert
    expect(response.statusCode).toEqual(404);
    expect(response.json()).toMatchSnapshot();
  });

  test('Create mock', async () => {
    // Arrange
    const serviceMockOptions = createTestServiceMockOptions();
    const id = uuidV4();
    const mockManagementService = app.get(MockManagementService);
    jest.spyOn(mockManagementService, 'createMock').mockReturnValueOnce(id);

    // Act
    const response = await app.inject({
      url: '/',
      method: 'POST',
      payload: serviceMockOptions,
    });

    // Assert
    expect(response.statusCode).toEqual(201);
    expect(response.json()).toEqual({ id });
  });

  test('Create mock with port that is already in use', async () => {
    // Arrange
    const serviceMockOptions = createTestServiceMockOptions();
    const mockManagementService = app.get(MockManagementService);
    jest.spyOn(mockManagementService, 'createMock').mockImplementation(() => {
      throw new Error(Constants.ERR_MOCK_PORT_ALREADY_IN_USE);
    });

    // Act
    const response = await app.inject({
      url: '/',
      method: 'POST',
      payload: serviceMockOptions,
    });

    // Assert
    expect(response.statusCode).toEqual(400);
    expect(response.json()).toMatchSnapshot();
  });

  test('Delete mock', async () => {
    // Arrange
    const serviceMock = createTestServiceMock();
    const mockManagementService = app.get(MockManagementService);
    jest.spyOn(mockManagementService, 'deleteMock').mockReturnValueOnce(serviceMock);

    // Act
    const response = await app.inject({
      url: `/${serviceMock.id}`,
      method: 'DELETE',
    });

    // Assert
    expect(response.statusCode).toEqual(200);
    expect(response.json()).toEqual(serviceMock.info);
  });
});
