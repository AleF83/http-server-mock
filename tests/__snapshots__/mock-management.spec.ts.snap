// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mock Management Create mock with port that is already in use 1`] = `
Object {
  "error": "Bad Request",
  "message": "The port is already in use",
  "statusCode": 400,
}
`;

exports[`Mock Management Get mock by nonexistent id 1`] = `
Object {
  "error": "Not Found",
  "message": "The mock with giver id doesn't exist",
  "statusCode": 404,
}
`;
